crates/api/src/externals.rs

```diff
@@ -451,36 +615,96 @@ impl Memory {
         }
     }
 
-    /// Returns a mutable slice the current memory.
+    /// Returns this memory as a slice view that can be read natively in Rust.
+    ///
+    /// # Safety
+    ///
+    /// This is an unsafe operation because there is no guarantee that the
+    /// following operations do not happen concurrently while the slice is in
+    /// use:
+    ///
+    /// * Data could be modified by calling into a wasm module.
+    /// * Memory could be relocated through growth by calling into a wasm
+    ///   module.
+    /// * When threads are supported, non-atomic reads will race with other
+    ///   writes.
+    ///
+    /// Extreme care need be taken when the data of a `Memory` is read. The
+    /// above invariants all need to be upheld at a bare minimum, and in
+    /// general you'll need to ensure that while you're looking at slice you're
+    /// the only one who can possibly look at the slice and read/write it.
+    ///
+    /// Be sure to keep in mind that `Memory` is reference counted, meaning
+    /// that there may be other users of this `Memory` instance elsewhere in
+    /// your program. Additionally `Memory` can be shared and used in any number
+    /// of wasm instances, so calling any wasm code should be considered
+    /// dangerous while you're holding a slice of memory.
+    pub unsafe fn data_unchecked(&self) -> &[u8] {
+        self.data_unchecked_mut()
+    }
+
+    /// Returns this memory as a slice view that can be read and written
+    /// natively in Rust.
+    ///
     /// # Safety
-    /// Marked unsafe due to posibility that wasmtime can resize internal memory
-    /// from other threads.
-    pub unsafe fn data(&self) -> &mut [u8] {
+    ///
+    /// All of the same safety caveats of [`Memory::data_unchecked`] apply
+    /// here, doubly so because this is returning a mutable slice! As a
+    /// double-extra reminder, remember that `Memory` is reference counted, so
+    /// you can very easily acquire two mutable slices by simply calling this
+    /// function twice. Extreme caution should be used when using this method,
+    /// and in general you probably want to result to unsafe accessors and the
+    /// `data` methods below.
+    pub unsafe fn data_unchecked_mut(&self) -> &mut [u8] {
         let definition = &*self.wasmtime_memory_definition();
         slice::from_raw_parts_mut(definition.base, definition.current_length)
     }
 
+    /// Returns the base pointer, in the host's address space, that the memory
+    /// is located at.
+    ///
+    /// When reading and manipulating memory be sure to read up on the caveats
+    /// of [`Memory::data_unchecked`] to make sure that you can safely
+    /// read/write the memory.
```

Comment:
Can we note something that will avoid the problem mentioned in https://www.reddit.com/r/rust/comments/e18jcq/how_to_read_and_write_to_memory_in_wasmtime/ ?

Reply:
I tried to write out how there's lots of caveats and the pointer can change a lot on `data_unchecked`, but do you have an idea of what sort of wording should be added there?

Reply:
Sorry, I don't. But the above reddit problem associated with memory moving in different location after grow operation, which is not connected with reference counting.

Reply:
Hm ok, I've attempted to address that with "Memory could be relocated through growth by calling into a wasm module". We can probably continue to wordsmith as issues come up though.

Reply:
FWIW, I made a similar warning in the .NET API around [`Memory.Span`](https://github.com/bytecodealliance/wasmtime/blob/master/crates/misc/dotnet/src/Memory.cs#L52-L57).